File: language/python/engineering-patterns/engineering_patterns/di_container.py
"""
手写 DI Container。

考点：
- 注册类型/工厂到容器。
- 按签名自动解析依赖（constructor injection）。
- 支持单例和 transient 生命周期。
"""

from __future__ import annotations

import inspect
from collections.abc import Callable
from enum import Enum, auto
from typing import Any, TypeVar

T = TypeVar("T")


class Lifetime(Enum):
    TRANSIENT = auto()
    SINGLETON = auto()


class DIContainer:
    """Minimal dependency injection container."""

    def __init__(self) -> None:
        self._registrations: dict[type, tuple[Lifetime, Callable[[], Any]]] = {}
        self._singletons: dict[type, Any] = {}

    def register(
        self,
        interface: type[T],
        factory: Callable[..., T] | type[T],
        *,
        lifetime: Lifetime = Lifetime.TRANSIENT,
    ) -> None:
        def factory_wrapper() -> T:
            return self._construct(factory)

        self._registrations[interface] = (lifetime, factory_wrapper)

    def resolve(self, interface: type[T]) -> T:
        if interface not in self._registrations:
            # Auto-registration: treat interface as a concrete type.
            return self._construct(interface)

        lifetime, factory = self._registrations[interface]
        if lifetime == Lifetime.SINGLETON:
            if interface not in self._singletons:
                self._singletons[interface] = factory()
            return self._singletons[interface]

        return factory()

    def _construct(self, cls: Callable[..., T]) -> T:
        if not inspect.isclass(cls) and not callable(cls):
            raise TypeError(f"Cannot construct {cls!r}")

        try:
            signature = inspect.signature(cls)
        except ValueError:
            return cls()

        kwargs: dict[str, Any] = {}
        for name, param in signature.parameters.items():
            if param.default is not inspect.Parameter.empty:
                continue
            annotation = param.annotation
            if annotation is inspect.Parameter.empty:
                raise ValueError(f"Cannot resolve parameter {name!r} of {cls!r}")
            kwargs[name] = self.resolve(annotation)

        return cls(**kwargs)


if __name__ == "__main__":
    class Database:
        def __init__(self, url: str = "sqlite://") -> None:
            self.url = url

    class UserService:
        def __init__(self, db: Database) -> None:
            self.db = db

    container = DIContainer()
    container.register(Database, Database, lifetime=Lifetime.SINGLETON)
    container.register(UserService, UserService)

    service = container.resolve(UserService)
    print(service.db.url)

File: language/typescript/engineering-patterns/diContainer.ts
/**
 * 手写 DI Container
 *
 * 考点：
 * - 注册类型/工厂。
 * - 按构造参数自动解析依赖。
 * - 支持 singleton / transient 生命周期。
 */

export type Constructor<T> = new (...args: unknown[]) => T;
export type Factory<T> = () => T;

export class DIContainer {
  private registrations = new Map<symbol | string, { factory: Factory<unknown>; singleton: boolean }>();
  private singletons = new Map<symbol | string, unknown>();

  register<T>(
    token: symbol | string,
    factory: Factory<T>,
    options: { singleton?: boolean } = {}
  ): void {
    this.registrations.set(token, { factory, singleton: options.singleton ?? false });
  }

  resolve<T>(token: symbol | string): T {
    const registration = this.registrations.get(token);
    if (!registration) {
      throw new Error(`No registration for token ${String(token)}`);
    }

    if (registration.singleton) {
      if (!this.singletons.has(token)) {
        this.singletons.set(token, registration.factory());
      }
      return this.singletons.get(token) as T;
    }

    return registration.factory() as T;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  class Database {
    url = "postgres://localhost";
  }
  class Service {
    constructor(public db: Database) {}
  }

  const container = new DIContainer();
  container.register("db", () => new Database(), { singleton: true });
  container.register("service", () => new Service(container.resolve("db")));

  const service = container.resolve<Service>("service");
  console.log(service.db.url);
}
